### ISSUE TITLE
Simplify TreeVAE code

### ISSUE THREAD

**Issue by romain-lopez** [khalilouardini/Cassiopeia_Transcriptome#2]
Hi Khalil, 

let's do the following to make the code more readable:

- [ ] Remove inheritance to [VAE class](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L21)
- [ ] Remove [batch parameter](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L50)
- [ ] Keep only support for [gene specific dispersion](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L56)
- [ ] Remove [log-variational parameter](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L57), always set to True
- [ ] Refactor [this](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L78) function
- [ ] These two [history lists](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L132) can be created on the fly, but are misleading in the constructor
- [ ] Use torch.zeros(dtype=) in place of [this](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/bf7c5184f082dc4d5958eb63e3999e4f63a890a3/scvi/external/models/treevae.py#L150) (or torch.zeros_like())
- [ ] Remove [this function](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L327) and factor the content in the forward function
- [ ] Remove ZINB support, only keep Poisson and NB. It will make the code easier to read
- [ ] Simplify signature of [the inference function](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L350) and remove the following arguments (batch_index=None, y=None, transform_batch=None)
- [ ] Remove [this line](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L363), not needed because we use observed library size 
- [ ] Remove z_transformation [stuff](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L370)
- [ ] Factor out all the inference code [that's not helping](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L390) 
- [ ] black your code and use isort
- [ ] Refactor the other files to remove all the additional code for zero inflation, for example [that one](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/inference/tree_inference.py)
- [ ] This is [unecessary](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/inference/tree_inference.py#L176)
- [ ] You don't need the inference function to calculate the (library size)[https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/inference/tree_inference.py#L322]
- [ ] Calculate the beginning before logging the input, otherwise it's scary for the reader (https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L382)

About the model itself:
- [ ] Replace your [FCLayers](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/models/modules.py#L20) by the one from the (new codebase)[https://github.com/YosefLab/scvi-tools/blob/master/scvi/nn/_base_components.py#L20]
- [ ] Build your own decoder in the constructor instead of using the [one from scVI](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L386). Your case is simpler, use just a neural net from z to the normalized mean of the negative binomial.
- [ ] Use layer norm instead of batch-norm in your non-linear experiments. e.g., refer to [this code](https://github.com/YosefLab/scvi-tools/blob/master/scvi/module/_vaec.py#L81)
- [ ] I don't like this idea of using the [LDecoder inside of our class](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L121). Let's aim at making the code more independent and not use this. Explicitly build a decoder in pytorch with matrix multiplication (the LDecoder has batch-norm and other crazy stuff in there)
- [ ] There are some dangerous inadequcies that come from commenting / uncommenting. For example [here](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/inference/tree_inference.py#L146). Let's also remove calls to pdb etc.. in this codebase and avoid committing those. Also, [here](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L382) it seems that the MP is not used for training. You can add options to the class instance such as "use_MP" for altering this behavior.

Misc.:
- [ ] do we just have the likelihood of a Gaussian vector [here](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/models/treevae.py#L303), if yes let's add a comment with which likelihood it is exactly (i.e., xxx \sim Normal(yyy, zzz))
- [ ] we must be careful about [this](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/inference/tree_inference.py#L362). I think the naive MC integral would calculate the mean after the MP. Not the MP of the mean. I this the latter is incorrect.
- [ ] [This](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/inference/tree_inference.py#L362) is inefficient? Use .sample(1000,) from torch ?
- [ ] If you have [this code](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/inference/tree_inference.py#L421), you might as well delete the self.variance_encoder in the TreeVAE class.
- [ ] I'd really want the [cross-validation](https://github.com/khalilouardini/Cassiopeia_Transcriptome/blob/master/scvi/external/inference/tree_inference.py#L488) code to work for reporting results

**Comment by romain-lopez** [comment]
Let me know if you have any questions!
